src/util/codec/bytes.rs

```diff
@@ -199,11 +210,83 @@ pub fn decode_bytes(data: &mut BytesSlice, desc: bool) -> Result<Vec<u8>> {
                 *k = !*k;
             }
         }
+        // data will point to following unencoded bytes, maybe timestamp
         *data = &data[offset..];
         return Ok(key);
     }
 }
 
+/// `decode_bytes_in_place` decodes bytes which is encoded by `encoded_bytes` before just in place without malloc.
+/// Please use this instead of `decode_bytes` if possible.
+pub fn decode_bytes_in_place(mut data: Vec<u8>, desc: bool) -> Result<Vec<u8>> {
+    let mut write_offset = 0;
+    let mut read_offset = 0;
+
+    let chunk_len = ENC_GROUP_SIZE + 1;
+    loop {
+        // everytime make ENC_GROUP_SIZE + 1 elements as a decode unit
+        let next_offset = read_offset + chunk_len;
+        if next_offset > data.len() {
+            return Err(Error::unexpected_eof());
+        };
+
+        // the last byte in decode unit is for marker which indicates pad size 
+        let marker = data[next_offset - 1];
+        let pad_size = if desc {
+            marker as usize
+        } else {
+            (ENC_MARKER - marker) as usize
+        };
+
+        // no padding, just move 8 bytes to write_offset
+        if pad_size == 0 {
+            unsafe {
+                // it is semantically equivalent to C's memmove()
+                // and the src and dest may overlap
+                ptr::copy(
```

Comment:
If read_offset == write_offset, don't need to copy.

Reply:
I've checked glibc and It seems it didn't skip the condition where `dst == src`. I agree that you do not copy when read_offset == write_offset.